e2e/test/MessageReceiveE2ETests.cs

```diff
@@ -165,16 +335,126 @@ private async Task ReceiveSingleMessage(TestDeviceType type, Client.TransportTyp
 
+                if (p1Value == prop2.Value)
+                {
+                    VerifyMessage(messageId1, payload1, p1Value, receivedMessage2);
+                    await deviceClient.CompleteAsync(receivedMessage2).ConfigureAwait(false);
+                }
+                else if (p1Value == prop3.Value)
+                {
+                    VerifyMessage(messageId1, payload1, p1Value, receivedMessage3);
+                    await deviceClient.CompleteAsync(receivedMessage3).ConfigureAwait(false);
+                }
+                else
+                {
+                    throw new InvalidOperationException(" Did not receive Abandoned message");
+                }
+    
                 await deviceClient.CloseAsync().ConfigureAwait(false);
                 await serviceClient.CloseAsync().ConfigureAwait(false);
             }
         }
 
+        public static async Task SendReceiveMultipleMessages(TestDeviceType type, Client.TransportType transport)
+        {
+            TestDevice testDevice = await TestDevice.GetTestDeviceAsync(DevicePrefix, type).ConfigureAwait(false);
+            CancellationTokenSource cts = new CancellationTokenSource(TimeSpan.FromMinutes(5));
+
+            using (DeviceClient deviceClient = testDevice.CreateDeviceClient(transport))
+            using (ServiceClient serviceClient = ServiceClient.CreateFromConnectionString(Configuration.IoTHub.ConnectionString))
+            {
+                await deviceClient.OpenAsync(cts.Token).ConfigureAwait(false);
+
+                if (transport == Client.TransportType.Mqtt_Tcp_Only ||
+                    transport == Client.TransportType.Mqtt_WebSocket_Only)
+                {
+                    // Dummy ReceiveAsync to ensure mqtt subscription registration before SendAsync() is called on service client.
+                    CancellationTokenSource cts20 = new CancellationTokenSource(TimeSpan.FromSeconds(20));
+                    await deviceClient.ReceiveAsync(cts20.Token).ConfigureAwait(false);
+                }
+                
+                
+                await serviceClient.OpenAsync().ConfigureAwait(false);
+                await serviceClient.PurgeMessageQueueAsync(testDevice.Id).ConfigureAwait(false);
+
+                int numMessages = 0;
+                CancellationTokenSource cts2 = new CancellationTokenSource(TimeSpan.FromSeconds(20));
+
+                /* DeviceQueue on service holds at max 50 messages which are not acked i.e. complete/reject/abandon 
+                 * from device Client, for receiving the 51st message without acking previous exception should be raised */
+                try
+                {
+                    for (int i = 0; i < 51; i++)
+                    {
+                        (Message msg1, string messageId1, string payload1, string p1Value) = ComposeC2DTestMessage();
+                        msg1.ExpiryTimeUtc = new DateTime(2019, 10, 24);
```

Comment:
Hardcoding the expiry time to 10/24/2019 will cause messages to be marked as expired (in the future). Use something like (CurrentTime + 1 hour) instead.

Reply:
agreed, good catch. will do